integrations/access/datadog/client.go

```diff
@@ -0,0 +1,236 @@
+	return func(_ *resty.Client, resp *resty.Response) error {
+		log := logger.Get(resp.Request.Context())
+
+		if sink != nil {
+			status := common.StatusFromStatusCode(resp.StatusCode())
+			// No usable context in scope, use background with a reasonable timeout
+			ctx, cancel := context.WithTimeout(context.Background(), statusEmitTimeout)
+			defer cancel()
+
+			if err := sink.Emit(ctx, status); err != nil {
+				log.WithError(err).Errorf("Error while emitting Datadog Incident Management plugin status: %v", err)
+			}
+		}
+
+		if resp.IsError() {
+			var details string
+			switch result := resp.Error().(type) {
+			case *ErrorResult:
+				details = fmt.Sprintf("http error code=%v, errors=[%v]", resp.StatusCode(), strings.Join(result.Errors, ", "))
+			default:
+				details = fmt.Sprintf("unknown error result %#v", result)
+			}
+			return trace.Errorf(details)
+		}
+		return nil
+	}
+}
+
+// CheckHealth pings Datadog and ensures required permissions.
+func (d *Datadog) CheckHealth(ctx context.Context) error {
+	var result PermissionsBody
+	_, err := d.client.NewRequest().
+		SetContext(ctx).
+		SetResult(&result).
+		Get("permissions")
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, permission := range result.Data {
+		if permission.Attributes.Name == IncidentWritePermissions && permission.Attributes.Restricted {
+			return trace.AccessDenied("missing incident_write permissions")
+		}
+	}
+	return nil
+}
+
+// Create Incident creates a new Datadog incident.
+func (d *Datadog) CreateIncident(ctx context.Context, summary string, recipients []common.Recipient, reqData pd.AccessRequestData) (IncidentsData, error) {
+	teams := make([]string, 0, len(recipients))
+	emails := make([]NotificationHandle, 0, len(recipients))
+
+	for _, recipient := range recipients {
+		switch recipient.Kind {
+		case common.RecipientKindTeam:
+			teams = append(teams, recipient.Name)
+		case common.RecipientKindEmail:
+			emails = append(emails, NotificationHandle{Handle: recipient.Name})
+		}
+	}
+
+	body := IncidentsBody{
+		Data: IncidentsData{
+			Metadata: Metadata{
+				Type: "incidents",
+			},
+			Attributes: IncidentsAttributes{
+				Title: fmt.Sprintf("Access request from %s", reqData.User),
+				Fields: IncidentsFields{
+					Summary: &StringField{
+						Type:  "textbox",
+						Value: summary,
+					},
+					State: &StringField{
+						Type:  "dropdown",
+						Value: "active",
+					},
+					DetectionMethod: &StringField{
+						Type:  "dropdown",
+						Value: "employee",
+					},
+					Severity: &StringField{
+						Type:  "dropdown",
+						Value: d.Severity,
+					},
+					RootCause: &StringField{
+						Type:  "textbox",
+						Value: reqData.RequestReason,
+					},
+					Teams: &StringSliceField{
+						Type:  "multiselect",
+						Value: teams,
+					},
+				},
+				NotificationHandles: emails,
+			},
+		},
+	}
+	var result IncidentsBody
+	_, err := d.client.NewRequest().
```

Comment:
Historical context: for azure/msteams we did not use the official API because the SDK depends on the entire universe and it took 30 min + to build the plugin (I tried doing it but gave up).

For other implementations this was done by the martians and they likely walked the easiest/quickest pah.

I don't have an opinion on whether we should use or not their go client. I think the most dangerous things with building our own client are the authentication flow and the failure/retry logic. If those are trivial and the go client adds many deps, rolling our own client makes sense.

In this case maybe drop a comment saying why we decided to do it ourselves.

Reply:
I'll leave a todo to migrate to the official datadog package if the need arises.